src/main/java/org/cobweb/cobweb2/plugins/disease/DiseaseMutator.java

```diff
@@ -1,251 +1,280 @@
+			// TODO: Maybe we can determine the effectiveness by the vaccinated agent types
+			vaccinate(bumpee, params.agentParams[tr].vaccineEffectiveness);
+		}
+
+		// Modify the logics here to not remove vaccination after being healed
+		if (params.agentParams[tr].healer && isSick(bumpee) && params.agentParams[tr].canHeal[te]) {
+			if (simulation.getRandom().nextFloat() < params.agentParams[tr].healerEffectiveness) {
+				heal(bumpee);
+			}
+		}
+
+		if (!isSick(bumper))
+			return;
+
+		if(isWearingPPE(bumpee)
+				&& simulation.getRandom().nextFloat() < params.agentParams[te].ppeEffectiveness)
+			return;
+
+		if(isWearingPPE(bumper)
+				&& simulation.getRandom().nextFloat() < params.agentParams[tr].ppeEffectiveness)
+			return;
+
+		if (isVaccinated(bumpee)
+				&& simulation.getRandom().nextFloat() < getAgentState(bumpee).vaccineEffectiveness)
+			return;
+
+		if (isSick(bumpee))
+			return;
+
+		if (params.agentParams[tr].transmitTo[te]) {
+			contactTransmit(bumpee, params.agentParams[te].contactTransmitRate);
+		}
+	}
+
+	public void unSick(Agent agent) {
+		removeAgentState(agent);
+		sickCount[agent.getType()]--;
+	}
+
+	public boolean isSick(Agent agent) {
+		return hasAgentState(agent) && getAgentState(agent).sick;
+	}
+
+	public boolean isVaccinated(Agent agent) {
+		return hasAgentState(agent) && getAgentState(agent).vaccinated;
+	}
+
+	public boolean isWearingPPE(Agent agent) {
+		return hasAgentState(agent) && getAgentState(agent).wearingPPE;
+	}
+
+	// This method is for vaccinators to vaccinate agents
+	public void vaccinate(Agent bumpee, float effectiveness) {
+		DiseaseAgentParams agentParams = params.agentParams[bumpee.getType()];
+		setAgentState(bumpee, new DiseaseState(agentParams, false, false, effectiveness, isWearingPPE(bumpee), agentParams.ppeEffectiveness));
+	}
+
+	// This method is for users to specifically vaccinate an individual agent
+	public void vaccinate(Agent agent) {
+		DiseaseAgentParams agentParams = params.agentParams[agent.getType()];
+		setAgentState(agent, new DiseaseState(agentParams, false, false, 0.0f, isWearingPPE(agent), agentParams.ppeEffectiveness));
+	}
+
+	public void deVaccinate(Agent agent) {
+		DiseaseAgentParams agentParams = params.agentParams[agent.getType()];
+		setAgentState(agent, new DiseaseState(agentParams, false, false, 0.0f, isWearingPPE(agent), agentParams.ppeEffectiveness));
+	}
+
+	public void heal(Agent agent) {
+		if (isVaccinated(agent)) {
+			DiseaseAgentParams agentParams = params.agentParams[agent.getType()];
+			setAgentState(agent, new DiseaseState(agentParams, false, true, agentParams.vaccineEffectiveness, isWearingPPE(agent), agentParams.ppeEffectiveness));
+			sickCount[agent.getType()]--;
+		} else {
+			unSick(agent);
+		}
+	}
+
+	public void equipPPE(Agent agent) {
+		DiseaseAgentParams agentParams = params.agentParams[agent.getType()];
+		setAgentState(agent, new DiseaseState(agentParams, isSick(agent), isVaccinated(agent), true, simulation.getTime()));
```

Comment:
This constructor of DiseaseState is used when an agent gets sick.
You need to use the other constructor.